File: src/main/java/io/github/gdpl2112/service/FileService.java
package io.github.gdpl2112.service;

import io.github.gdpl2112.model.User;
import io.github.gdpl2112.model.UserFile;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

@Service
public class FileService {

    @Value("${file.upload-dir}")
    private String uploadDir;

    /**
     * 获取用户文件列表
     *
     * @param user 用户对象
     * @return 文件列表
     */
    public List<UserFile> getUserFiles(User user) {
        List<UserFile> files = new ArrayList<>();
        if (user == null) {
            return files;
        }

        // 构建用户文件夹路径（可以根据实际需求调整）
        String userFolderPath = user.getPathDir(uploadDir); // 这里简化处理，实际可能需要根据用户ID创建独立文件夹
        File userFolder = new File(userFolderPath);

        if (userFolder.exists() && userFolder.isDirectory()) {
            collectFiles(userFolder, files, "");
        }

        return files;
    }

    /**
     * 递归收集文件
     *
     * @param folder   文件夹
     * @param files    文件列表
     * @param basePath 基础路径
     */
    private void collectFiles(File folder, List<UserFile> files, String basePath) {
        File[] fileList = folder.listFiles();
        if (fileList != null) {
            for (File file : fileList) {
                if (file.isFile()) {
                    UserFile userFile = new UserFile();
                    userFile.setName(file.getName());
                    userFile.setSize(file.length());
                    userFile.setPath(basePath.isEmpty() ? file.getName() : basePath + "/" + file.getName());
                    files.add(userFile);
                } else if (file.isDirectory()) {
                    // 递归处理子文件夹
                    String newBasePath = basePath.isEmpty() ? file.getName() : basePath + "/" + file.getName();
                    collectFiles(file, files, newBasePath);
                }
            }
        }
    }

    /**
     * 计算文件夹大小
     *
     * @param folder 文件夹
     * @return 文件夹大小（字节）
     */
    public long calculateFolderSize(File folder) {
        long size = 0;
        File[] files = folder.listFiles();
        if (files != null) {
            for (File file : files) {
                if (file.isFile()) {
                    size += file.length();
                } else if (file.isDirectory()) {
                    size += calculateFolderSize(file);
                }
            }
        }
        return size;
    }

    public long calculateFolderSize(User user) {
        // 构建用户文件夹路径（可以根据实际需求调整）
        String userFolderPath = user.getPathDir(uploadDir); // 这里简化处理，实际可能需要根据用户ID创建独立文件夹
        File userFolder = new File(userFolderPath);
        if (userFolder.exists() && userFolder.isDirectory()) {
            return calculateFolderSize(userFolder);
        }
        return 0;
    }

}

File: gofiles/service/FileService.go
package service

import (
	"os"
	"path/filepath"

	"github.com/gdpl2112/files-server/model"
)

// FileService 文件服务
type FileService struct {
	UploadDir string
}

// NewFileService 创建文件服务实例
func NewFileService(uploadDir string) *FileService {
	return &FileService{
		UploadDir: uploadDir,
	}
}

// GetUserFiles 获取用户文件列表
func (s *FileService) GetUserFiles(user *model.User) []*model.UserFile {
	files := make([]*model.UserFile, 0)
	if user == nil {
		return files
	}

	userFolderPath := user.GetPathDir(s.UploadDir)
	if info, err := os.Stat(userFolderPath); err == nil && info.IsDir() {
		s.collectFiles(userFolderPath, &files, "")
	}

	return files
}

// collectFiles 递归收集文件
func (s *FileService) collectFiles(folderPath string, files *[]*model.UserFile, basePath string) {
	entries, err := os.ReadDir(folderPath)
	if err != nil {
		return
	}

	for _, entry := range entries {
		entryPath := filepath.Join(folderPath, entry.Name())
		if entry.IsDir() {
			newBasePath := basePath
			if newBasePath != "" {
				newBasePath += "/"
			}
			newBasePath += entry.Name()
			s.collectFiles(entryPath, files, newBasePath)
		} else {
			fileInfo, err := entry.Info()
			if err != nil {
				continue
			}

			userFile := &model.UserFile{
				Name: entry.Name(),
				Size: fileInfo.Size(),
			}

			if basePath != "" {
				userFile.Path = basePath + "/" + entry.Name()
			} else {
				userFile.Path = entry.Name()
			}

			*files = append(*files, userFile)
		}
	}
}

// CalculateFolderSize 计算文件夹大小
func (s *FileService) CalculateFolderSize(folderPath string) int64 {
	var size int64

	err := filepath.Walk(folderPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() {
			size += info.Size()
		}
		return nil
	})

	if err != nil {
		return 0
	}

	return size
}

// CalculateUserFolderSize 计算用户文件夹大小
func (s *FileService) CalculateUserFolderSize(user *model.User) int64 {
	if user == nil {
		return 0
	}

	userFolderPath := user.GetPathDir(s.UploadDir)
	if info, err := os.Stat(userFolderPath); err == nil && info.IsDir() {
		return s.CalculateFolderSize(userFolderPath)
	}

	return 0
}
